src/server/model/modelGraph.py

```diff
@@ -0,0 +1,117 @@
+        self._node = node
+        self._parent = parent
+        if self._parent is not None:
+            self._parent.add_child(self)
+        self._children = {}
+        if children is not None:
+            for child in children:
+                self.add_child(child)
+
+    def add_child(self, node):
+        if isinstance(node, GraphNode):
+            self._children[node.name] = node
+        else:
+            self._children[node] = GraphNode(node, self)
+
+    def remove_child(self, node):
+        if node.name in list(self._children):
+            del self._children[node.name]
+
+    def get_child(self, name):
+        if name in list(self._children):
+            return self._children[name]
+        else:
+            return None
+
+    def get_children(self):
+        return self._children
+
+    def get_parent(self):
+        return self._parent
+
+    def get_node(self):
+        return self._node
+
+    def set_parent(self, parent):
+        if parent is not None:
+            if isinstance(parent, GraphNode) is False:
+                raise Exception("Parent must be a GraphNode")
+            if self._parent is not None:
+                self._parent.remove_child(self)
+            parent.add_child(self)
+        else:
+            if self._parent is not None:
+                self._parent.remove_child(self)
+        self._parent = parent
+
+    def set_node(self, node):
+        self._node = node
+
+
+class ModelGraph:
+    def __init__(self):
+        self._graph = {}
+        self._nodes_map = []
+
+    def add_node(self, node, parent=None):
+        # ensure we are not duplicating the node
+        node_found = self.find_node(node)
+        if node_found is None:
+            node_found = node if isinstance(node, GraphNode) else GraphNode(node)
+            if parent is not None:
+                parent_found = self.find_node(parent)
+                if parent_found is None:
+                    parent_found = GraphNode(parent)
+                    self.add_node(parent)
+                node_found.set_parent(parent_found)
+            else:
+                self._graph[node.name] = node_found
+            return True
+        else:
+            return False
+
+    def move_node(self, node, parent):
+        node_found = self.find_node(node)
+        parent_found = self.find_node(parent)
+        if parent_found is None:
+            self.add_node(parent)
+            parent_found = self.find_node(parent)
+        if node_found is not None:
+            if node_found.get_parent() is None and self._graph[node.name] is not None:
+                del self._graph[node.name]
+            node_found.set_parent(parent_found)
+        else:
+            self.add_node(node, parent_found)
+
+    def find_node(self, node):
+        for graphNode in self.get_all_nodes():
+            if node.name == graphNode.name:
+                return graphNode
+        return None
+
+    def get_node_children(self, node):
```

Comment:
Inspired by your yield from below (which I never seen before, so thank you for that) I think we can do the same here:

```
def get_node_children(self, node):
    for child in node.get_children().values():
        yield child
        yield from self.get_node_children(child)
```
Take it with a bit of salt because I just learned about yield from :smile: